presto-main/src/main/java/com/facebook/presto/type/TimestampWithTimeZoneOperators.java

```diff
@@ -137,9 +138,15 @@ public static long castToTimeWithTimeZone(@SqlType(StandardTypes.TIMESTAMP_WITH_
 
     @ScalarOperator(CAST)
     @SqlType(StandardTypes.TIMESTAMP)
-    public static long castToTimestamp(@SqlType(StandardTypes.TIMESTAMP_WITH_TIME_ZONE) long value)
+    public static long castToTimestamp(ConnectorSession session, @SqlType(StandardTypes.TIMESTAMP_WITH_TIME_ZONE) long value)
     {
-        return unpackMillisUtc(value);
+        if (session.isLegacyTimestamp()) {
+            return unpackMillisUtc(value);
+        }
+        else {
+            ISOChronology chronology = getChronology(unpackZoneKey(value));
+            return unpackMillisUtc(value) + chronology.getZone().getOffset(value);
```

Comment:
Nice catch! Thanks for that!
I fixed that and added test that will fail if someone does that again in this place.